### PR TITLE
Walker Environment

### PR DESCRIPTION
**Pull request by awjuliani**
💃 🕺 

walker

### CODE REVIEW

File: docs/Learning-Environment-Examples.md [hunk 1]
```diff
@@ -224,3 +224,22 @@ If you would like to contribute environments, please see our
...
     * Visual Observations: None
 * Reset Parameters: None
+
+## Walker
+
+![Walker](images/walker.png)
+
+* Set-up: A humanoid agent with multiple degrees of freedom.
+* Goal: The agents must move its body toward the goal direction as quickly as possible without falling.
+* Agents: The environment contains 11 agent linked to a single brain.
```
Comment (mmattar) [inline docs/Learning-Environment-Examples.md]: 11 agent*s*

Why 11 agents?

File: docs/Learning-Environment-Examples.md [hunk 2]
```diff
@@ -224,3 +224,22 @@ If you would like to contribute environments, please see our
...
+* Agent Reward Function (independent): 
+    * +0.03 times body velocity in the goal direction.
+    * +0.01 times head y position.
+    * +0.01 times body direction alignment with goal direction.
+    * -0.01 times head velocity difference from body velocity.
+    * Brains: One brain with the following observation/action space.
```
Comment (mmattar) [inline docs/Learning-Environment-Examples.md]: Un-indent

File: docs/Learning-Environment-Examples.md [hunk 3]
```diff
@@ -224,3 +224,22 @@ If you would like to contribute environments, please see our
         * Goalie: 4 corresponding to forward, backward, sideways movement.
     * Visual Observations: None
 * Reset Parameters: None
+
+## Walker
+
+![Walker](images/walker.png)
+
+* Set-up: A humanoid agent with multiple degrees of freedom.
```
Comment (mmattar) [inline docs/Learning-Environment-Examples.md]: List/describe the limbs.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WalkerAcademy : Academy
+{
+    public override void InitializeAcademy()
+    {
+        Monitor.verticalOffset = 1f;
+        Physics.defaultSolverIterations = 12;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs]: Can you comment on why the values are increased from the default.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        Physics.defaultSolverVelocityIterations = 12;
+    }
+
+    public override void AcademyReset()
+    {
+
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs]: Remove empty line

public override ...
{

}

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public override void AcademyStep()
+    {
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs]: Similarly, remove empty line.

File: unity-environment/ProjectSettings/ProjectSettings.asset [hunk 7]
```diff
@@ -535,15 +557,13 @@ PlayerSettings:
   webGLTemplate: APPLICATION:Default
   webGLAnalyzeBuildSize: 0
   webGLUseEmbeddedResources: 0
+  webGLUseWasm: 0
   webGLCompressionFormat: 1
-  webGLLinkerTarget: 0
   scriptingDefineSymbols:
-    1: 
+    1: ENABLE_TENSORFLOW
```
Comment (mmattar) [inline unity-environment/ProjectSettings/ProjectSettings.asset]: Default should be off.

File: unity-environment/ProjectSettings/EditorBuildSettings.asset [hunk 8]
```diff
@@ -4,4 +4,8 @@
 EditorBuildSettings:
   m_ObjectHideFlags: 0
   serializedVersion: 2
-  m_configObjects: {}
+  m_Scenes:
+  - enabled: 1
+    path: Assets/ML-Agents/Examples/New Demos/Walker/Scenes/Walker Using Conf Joint
```
Comment (mmattar) [inline unity-environment/ProjectSettings/EditorBuildSettings.asset]: Do not see the scene path in this PR.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WalkerGroundContact : MonoBehaviour
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Please add class-level description for this object. (What it does, when it should be used and any assumptions.)

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public int index;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Not used within this class.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    public WalkerAgent agent;
+    public bool touchingGround;
+    public bool penalizeOnContact;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Who should be setting this value and when would it be true vs false?

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    void Start()
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Please add 1-2 sentence method-level description

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        agent = transform.root.GetComponent<WalkerAgent>();
+    }
+
+    void OnCollisionEnter(Collision other)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Please add 1-2 sentence method-level description

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+            touchingGround = true;
+            if (penalizeOnContact)
+            {
+                agent.Done();
+                agent.SetReward(-1f);
+            }
+        }
+    }
+
+    void OnCollisionExit(Collision other)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Please add 1-2 sentence method-level description

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    public bool penalizeOnContact;
+
+    void Start()
+    {
+        agent = transform.root.GetComponent<WalkerAgent>();
+    }
+
+    void OnCollisionEnter(Collision other)
+    {
+        if (other.transform.CompareTag("ground"))
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: "ground" should be a global static variable somewhere.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+            {
+                agent.Done();
+                agent.SetReward(-1f);
+            }
+        }
+    }
+
+    void OnCollisionExit(Collision other)
+    {
+        if (other.transform.CompareTag("ground"))
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: "ground" should be a global static variable somewhere.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+        foreach (var item in bodyParts)
+        {
+            item.Key.position = item.Value.startingPos;
+            item.Key.rotation = item.Value.startingRot;
+            item.Value.rb.velocity = Vector3.zero;
+            item.Value.rb.angularVelocity = Vector3.zero;
+        }
+    }
+
+}
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Add new line at end of file

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+        
+        foreach (var item in bodyParts)
+        {
+            item.Key.position = item.Value.startingPos;
+            item.Key.rotation = item.Value.startingRot;
+            item.Value.rb.velocity = Vector3.zero;
+            item.Value.rb.angularVelocity = Vector3.zero;
+        }
+    }
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Remove empty line

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+    public Dictionary<Transform, BodyPart> bodyParts = new Dictionary<Transform, BodyPart>();
+
+    [System.Serializable]
+    public class BodyPart
+    {
+        public ConfigurableJoint joint;
+        public Rigidbody rb;
+        public Vector3 startingPos;
+        public Quaternion startingRot;
+        public WalkerGroundContact groundContactScript;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: I'd remove the "Script" suffix from variable name.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+        SetupBodyPart(thighR);
+        SetupBodyPart(shinR);
+        SetupBodyPart(footR);
+        SetupBodyPart(armL);
+        SetupBodyPart(forearmL);
+        SetupBodyPart(handL);
+        SetupBodyPart(armR);
+        SetupBodyPart(forearmR);
+        SetupBodyPart(handR);
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Remove empty line

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 21]
```diff
@@ -0,0 +1,189 @@
+﻿using System.Collections;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Please ensure that all code lines are within 100 characters.

File: python/unitytrainers/models.py [hunk 22]
```diff
@@ -118,7 +118,7 @@ def create_discrete_state_encoder(self, s_size, h_size, activation, num_layers):
     def create_new_obs(self, num_streams, h_size, num_layers):
         brain = self.brain
         s_size = brain.vector_observation_space_size * brain.num_stacked_vector_observations
-        if brain.vector_action_space_type == "continuous":
+        if brain.vector_action_space_size == "continuous":
```
Comment (mmattar) [inline python/unitytrainers/models.py]: Is this intentional?

File: unity-environment/Assets/ML-Agents/Examples/3DBall/Scripts/Ball3DAgent.cs [hunk 23]
```diff
@@ -19,15 +19,16 @@ public override void CollectObservations()
         AddVectorObs(gameObject.transform.rotation.x);
         AddVectorObs(ball.transform.position - gameObject.transform.position);
         AddVectorObs(ballRb.velocity);
+        AddVectorObs(ballRb.angularVelocity);
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/3DBall/Scripts/Ball3DAgent.cs]: Please move all 3DBall changes to its own PR. This PR ought to be limited to Walker environment changes.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+        SetupBodyPart(armL);
+        SetupBodyPart(forearmL);
+        SetupBodyPart(handL);
+        SetupBodyPart(armR);
+        SetupBodyPart(forearmR);
+        SetupBodyPart(handR);
+
+    }
+
+    public Quaternion GetJointRotation(ConfigurableJoint joint)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: - This can be a static method. Maybe move to a helper class, if we have one.
- Add method-level description

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        return (Quaternion.FromToRotation(joint.axis, joint.connectedBody.transform.rotation.eulerAngles));
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void BodyPartObservation(BodyPart bp)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: I'd recommend CollectObservationsBodyPart as the method name

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+
+    }
+
+    public Quaternion GetJointRotation(ConfigurableJoint joint)
+    {
+        return (Quaternion.FromToRotation(joint.axis, joint.connectedBody.transform.rotation.eulerAngles));
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Adds the observations for the provided body part.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+    public Transform armL;
+    public Transform forearmL;
+    public Transform handL;
+    public Transform armR;
+    public Transform forearmR;
+    public Transform handR;
+    public Dictionary<Transform, BodyPart> bodyParts = new Dictionary<Transform, BodyPart>();
+
+    [System.Serializable]
+    public class BodyPart
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Class-level description.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+    public Transform thighR;
+    public Transform shinR;
+    public Transform footR;
+    public Transform armL;
+    public Transform forearmL;
+    public Transform handL;
+    public Transform armR;
+    public Transform forearmR;
+    public Transform handR;
+    public Dictionary<Transform, BodyPart> bodyParts = new Dictionary<Transform, BodyPart>();
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Variable name `bodyParts` implies that this is a list of bodyPart objects, not a dictionary.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+        }
+    }
+
+    /// <summary>
+    /// Loop over body parts to add them to observation.
+    /// </summary>
+    public override void CollectObservations()
+    {
+        AddVectorObs(goalDirection);
+        foreach (var item in bodyParts)
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: foreach(var bodyPart in bodyParts.Values)

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+        {
+            BodyPartObservation(item.Value);
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: BodyPartObservation(bodyPart)

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+        {
+            positionSpring = ((strength + 1f) * 0.5f) * 10000f,
+            maximumForce = 250000f
+        };
+        bp.joint.slerpDrive = jd;
+    }
+
+    public override void AgentAction(float[] vectorAction, string textAction)
+    {
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: remove empty line

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+        AddVectorObs(goalDirection);
+        foreach (var item in bodyParts)
+        {
+            BodyPartObservation(item.Value);
+        }
+    }
+
+    /// <summary>
+    /// Apply torque according to defined goal angle and force.
+    /// </summary>
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: add param fields for method arguments

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 33]
```diff
@@ -0,0 +1,189 @@
...
+    public void SetNormalizedTargetRotation(BodyPart bp, float x, float y, float z, float strength)
+    {
+        x = (x + 1f) * 0.5f;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Comment on these three lines

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 34]
```diff
@@ -0,0 +1,189 @@
...
+        y = (y + 1f) * 0.5f;
+        z = (z + 1f) * 0.5f;
+
+        var xRot = Mathf.Lerp(bp.joint.lowAngularXLimit.limit, bp.joint.highAngularXLimit.limit, x);
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Comment on these three lines, particularly why the x-rotation is different from y and z.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 35]
```diff
@@ -0,0 +1,189 @@
...
+        var yRot = Mathf.Lerp(-bp.joint.angularYLimit.limit, bp.joint.angularYLimit.limit, y);
+        var zRot = Mathf.Lerp(-bp.joint.angularZLimit.limit, bp.joint.angularZLimit.limit, z);
+
+        bp.joint.targetRotation = Quaternion.Euler(xRot, yRot, zRot);
+        var jd = new JointDrive
+        {
+            positionSpring = ((strength + 1f) * 0.5f) * 10000f,
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: comment on 10000f variable - make it a class constant (?)

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 36]
```diff
@@ -0,0 +1,189 @@
...
+            maximumForce = 250000f
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: comment on 250000f variable - make it a class constant (?)

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 37]
```diff
@@ -0,0 +1,189 @@
...
+        // b. Rotation alignment with goal direction.
+        // c. Encourage head height.
+        // d. Discourage head movement.
+        AddReward(
+            + 0.03f * Vector3.Dot(goalDirection, bodyParts[hips].rb.velocity)
+            + 0.01f * Vector3.Dot(goalDirection, hips.forward)
+            + 0.01f * (head.position.y - hips.position.y)
+            - 0.01f * Vector3.Distance(bodyParts[head].rb.velocity, bodyParts[hips].rb.velocity)
+        );
+
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: remove empty line

File: unity-environment/Assets/ML-Agents/Examples/Bouncer/Scripts/BouncerAgent.cs [hunk 38]
```diff
@@ -1,4 +1,4 @@
-﻿using System.Collections;
+using System.Collections;
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Bouncer/Scripts/BouncerAgent.cs]: Is this normal, what is the difference ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgentConfigurable.cs [hunk 39]
```diff
@@ -1,4 +1,4 @@
-﻿using System.Collections;
+using System.Collections;
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgentConfigurable.cs]: There does not seem to be any difference.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs (hunk 9 above)
Comment (vincentpierre) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Yea, and we could reuse this in the crawler.

**Comment by vincentpierre** [comment]
Please remove the .asset files such as `unity-environment/ProjectSettings/EditorBuildSettings.asset`

File: docs/Learning-Environment-Examples.md (hunk 1 above)
Comment (awjuliani) [reply docs/Learning-Environment-Examples.md]: Somewhat arbitrarily choose. We want enough to speed up training, but not too many as to slow Unity/communication down. Also agents have to be spread apart, and if they are too far from the center of the scene, physics calculations become inaccurate.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs (hunk 4 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAcademy.cs]: @Hunter-Unity Could you comment on these?

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs (hunk 9 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Just changed it so that it is generic and re-usable, as well as commented.

File: unity-environment/ProjectSettings/ProjectSettings.asset (hunk 7 above)
Comment (awjuliani) [reply unity-environment/ProjectSettings/ProjectSettings.asset]: Removed this.

File: unity-environment/ProjectSettings/EditorBuildSettings.asset (hunk 8 above)
Comment (awjuliani) [reply unity-environment/ProjectSettings/EditorBuildSettings.asset]: Removed this.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs (hunk 11 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerGroundContact.cs]: Environment creator sets it, from the editor. True for body parts you don't want touching the ground; false for all others.

File: python/unitytrainers/models.py (hunk 22 above)
Comment (awjuliani) [reply python/unitytrainers/models.py]: No. It was a typo that has been fixed.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 21 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: My IDE default to 120. Are we going with 100 for everything?

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 34 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: @Hunter-Unity Can you explain this? I was wondering why X is different myself.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 35 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: @Hunter-Unity

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 36 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: @Hunter-Unity

**Comment by awjuliani** [comment]
@mmattar Adjusted the color scheme to give them blue (agent color) pants.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 40]
```diff
@@ -0,0 +1,191 @@
...
+        SetNormalizedTargetRotation(bodyParts[chest], vectorAction[0], vectorAction[1], vectorAction[2],
+            vectorAction[26]);
+        SetNormalizedTargetRotation(bodyParts[spine], vectorAction[3], vectorAction[4], vectorAction[5],
+            vectorAction[27]);
+
+        SetNormalizedTargetRotation(bodyParts[thighL], vectorAction[6], vectorAction[7], 0, vectorAction[28]);
+        SetNormalizedTargetRotation(bodyParts[shinL], vectorAction[8], 0, 0, vectorAction[29]);
+        SetNormalizedTargetRotation(bodyParts[footL], vectorAction[9], vectorAction[10], vectorAction[11],
+            vectorAction[30]);
+        SetNormalizedTargetRotation(bodyParts[thighR], vectorAction[12], vectorAction[13], 0, vectorAction[31]);
```
Comment (vladimiroster) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Why not encapsulate the transform, and the indices into vectorAction inside bodypart? The danger here (apart from having 2 different objects that refer to the same entity from different angles) is that mistakes in this method are easily introduced and can be difficult to catch.

File: unity-environment/Assets/ML-Agents/Examples/SharedAssets/Scripts/GroundContact.cs [hunk 41]
```diff
@@ -0,0 +1,51 @@
...
+    /// </summary>
+    void OnCollisionEnter(Collision other)
+    {
+        if (other.transform.CompareTag(Ground))
+        {
+            touchingGround = true;
+            if (penalizeOnContact)
+            {
+                agent.Done();
+                agent.SetReward(-1f);
```
[the remark this answers is not in the document]
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/SharedAssets/Scripts/GroundContact.cs]: Good idea. Just made the change.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 40 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: So the `BodyPart` class would contain the indices which correspond to `x`, `y`, `z`, and `strength` values, and here instead we do something like `SetNormalizedTargetRotation(bodyPart, vectorAction)`?

Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: So move `SetNormalizedTargetRotation` into the body part itself?

Comment (vladimiroster) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Yes. Move both the transform and SetNormalizedTargetRotation into BodyPart. Then this method will just enumerate all the body parts and call their SetNormalizedTargetRotation. The problem is that each BodyPart will need to know which indices to use, so the construction of the body parts will be more complicated.
Let's leave it as is for now, I'll try to think of a better way to match indices to body parts.

File: unity-environment/ProjectSettings/DynamicsManager.asset [hunk 42]
```diff
@@ -9,8 +9,8 @@ PhysicsManager:
   m_BounceThreshold: 2
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/DynamicsManager.asset]: Does this file need to be included in the PR ?

File: docs/Learning-Environment-Examples.md [hunk 43]
```diff
@@ -224,3 +224,22 @@ If you would like to contribute environments, please see our
...
     * Visual Observations: None
 * Reset Parameters: None
+
+## Walker
+
+![Walker](images/walker.png)
+
+* Set-up: A humanoid agent with 26 degrees of freedom. These correspond to articulation of the following body-parts: hips, chest, spine, head, thighs, shins, feets, arms, forearms and hands. 
+* Goal: The agents must move its body toward the goal direction as quickly as possible without falling.
+* Agents: The environment contains 11 agent linked to a single brain.
```
Comment (mmattar) [inline docs/Learning-Environment-Examples.md]: agent --> agents

File: unity-environment/Assets/ML-Agents/Examples/3DBall/3DBall.unity [hunk 44]
```diff
@@ -288,6 +288,18 @@ GameObject:
   m_PrefabParentObject: {fileID: 1536511242562482, guid: ff026d63a00abdc48ad6ddcff89aba04,
     type: 2}
   m_PrefabInternal: {fileID: 119733639}
+--- !u!114 &225807877
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/3DBall/3DBall.unity]: Should we remove this file too from the PR?

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 45]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void CollectObservationBodyPart(BodyPart bp)
+    {
+        var rb = bp.rb;
+        AddVectorObs(bp.groundContact.touchingGround ? 1 : 0); // Is this bp touching the ground
+        bp.groundContact.touchingGround = false;
```
Comment (mmattar) [inline unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: Why do we update `touchingGround`? A comment would be great.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 21 above)
Comment (mmattar) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: 80 for python (PEP 8 standard) / C++ and 100 for C#/Java.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 40 above)
Comment (mmattar) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: I like this suggestion @vladimiroster. Why would each body part need to be aware of `vectorAction` though?

The `SetNormalizedTargetRotation` would still accept 4 inputs (x,y,z,strength) and it is on the Agent class to retrieve those values from vectorAction/textAction and pass them accordingly.

File: unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 45 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: @Hunter-Unity ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgentConfigurable.cs (hunk 39 above)
Comment (xiaomaogy) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgentConfigurable.cs]: Usually this is due to line ending difference.
